File: leetgo/go/LCR 125.yong-liang-ge-zhan-shi-xian-dui-lie-lcof/solution.go
// Created by k at 2024/05/13 14:40
// leetgo: dev
// https://leetcode.cn/problems/yong-liang-ge-zhan-shi-xian-dui-lie-lcof/

package main

import (
	"bufio"
	"fmt"
	"os"

	. "github.com/j178/leetgo/testutils/go"
)

// @lc code=begin

type CQueue struct {
	head []int
	tail []int
}

func Constructor() CQueue {

	return CQueue{}
}

func (c *CQueue) AppendTail(value int) {
	c.tail = append(c.tail, value)
}

func (c *CQueue) DeleteHead() (ans int) {
	if len(c.tail) == 0 && len(c.head) == 0 {
		return -1
	}
	if len(c.head) == 0 {
		// c.head = append(c.head, c.tail...)
		// c.tail = c.tail[:0]
		c.head = c.tail
		c.tail = make([]int, 0)
	}
	ans = c.head[0]
	c.head = c.head[1:]

	return
}

// @lc code=end

func main() {
	stdin := bufio.NewReader(os.Stdin)
	ops := Deserialize[[]string](ReadLine(stdin))
	params := MustSplitArray(ReadLine(stdin))
	output := make([]string, 0, len(ops))
	output = append(output, "null")

	obj := Constructor()

	for i := 1; i < len(ops); i++ {
		switch ops[i] {
		case "appendTail":
			methodParams := MustSplitArray(params[i])
			value := Deserialize[int](methodParams[0])
			obj.AppendTail(value)
			output = append(output, "null")
		case "deleteHead":
			ans := Serialize(obj.DeleteHead())
			output = append(output, ans)
		}
	}
	fmt.Println("\noutput:", JoinArray(output))
}

File: leetgo/cpp/LCR 125.yong-liang-ge-zhan-shi-xian-dui-lie-lcof/solution.cpp
// Created by k at 2024/05/13 15:08
// leetgo: dev
// https://leetcode.cn/problems/yong-liang-ge-zhan-shi-xian-dui-lie-lcof/

#include <bits/stdc++.h>
#include <vector>
#include "LC_IO.h"
using namespace std;

// @lc code=begin

class CQueue {
  public:
    CQueue() {}

    void appendTail(int value) { tail.push_back(value); }

    int deleteHead()
    {
        if (head.empty() && tail.empty()) {
            return -1;
        }
        if (head.empty()) {
            head.insert(head.cbegin(), tail.cbegin(), tail.cend());
            tail.clear();
        }
        int ans = head.front();
        head.erase(head.begin());
        return ans;
    }
    vector<int> head, tail;
};

// @lc code=end

int main()
{
    ios_base::sync_with_stdio(false);
    stringstream out_stream;

    vector<string> method_names;
    LeetCodeIO::scan(cin, method_names);

    CQueue *obj;
    const unordered_map<string, function<void()>> methods = {
        {"CQueue",
         [&]() {
             cin.ignore();
             obj = new CQueue();
             out_stream << "null,";
         }},
        {"appendTail",
         [&]() {
             int value;
             LeetCodeIO::scan(cin, value);
             cin.ignore();
             obj->appendTail(value);
             out_stream << "null,";
         }},
        {"deleteHead",
         [&]() {
             cin.ignore();
             LeetCodeIO::print(out_stream, obj->deleteHead());
             out_stream << ',';
         }},
    };
    cin >> ws;
    out_stream << '[';
    for (auto &&method_name : method_names) {
        cin.ignore(2);
        methods.at(method_name)();
    }
    cin.ignore();
    out_stream.seekp(-1, ios_base::end);
    out_stream << ']';
    cout << "\noutput: " << out_stream.rdbuf() << endl;

    delete obj;
    return 0;
}
